04-frequent-scenarios.md

```diff
@@ -0,0 +1,288 @@
+@ResponseStatus(HttpStatus.NOT_MODIFIED)
+public void handleCarNotModified() {}
+```
+
+Additionally, if max-age is used (in combination with must-revalidate), the max-age response header must be set in this handler method as well. This causes the already cached result to be 'refreshed' with a new max-age value.
+
+### 400 / Validation error
+
+A validation error typically occurs when the input is not up to par with the service standard. The desired feedback consists of a field by field breakdown containing all the violated validation rules for every field. Any UI can then choose to decorate the field with the violated rules to inform the user.
+
+**Note:** that in an ongoing discussion, this error is sometimes generated under status code [422][422-definition]. The rationale behind using 422 instead of 400, is that 422 specifically speaks about syntactically correct, but semantically erroneous input. A validation error is just that.
+
+Two levels of validations are usually recognized:
+* validation which can be made in isolation; a typical example of this kind of validation is whether a zip code conforms to a nationally defined standard. 
+* validation which is made in context; a typical example of this kind of validation is to see if a similar record is already in place.  
+
+The first kind of validation lends itself very well to be implemented both in the UI (as standalone rules) and in the backend (in a similar vein). The second kind of validation is typically enforced by a service which has the capability to make calls to a repository layer.
+
+Both kinds can be made to fit in a generic validation framework. Java Bean Validation (aka JSR-303), is a decent and standardized way to deal with this. 
+
+If you want to be able to validate your input against standard validation rules, be sure to include the Hibernate validator (or a similar library):
+```xml
+<dependency>
+    <groupId>org.hibernate<
+    <artifactId>hibernate-validator</artifactId>
+    <version>5.1.3.Final</version>
+</dependency>
+```
+
+Make sure to configure the Bean Validator in your application configuration:
+
+```java
+@Bean
+public Validator validator() {
+    return new LocalValidatorFactoryBean();
+}
+```
+
+The Java SDK API offers a number of validation annotations in its javax.validation.constraints package. These annotations can be used to annotate input classes (such as forms):
+
+```java
+public class CarForm {
+
+    @NotEmpty @Size(max = 80)
+    public String name;
+}
+```
+
+It now becomes possible to further annotate your input classes in the controller methods to enforce validation on method entry:
+ 
+```java
+@RequestMapping(method = RequestMethod.POST)
+public Car createCar(@RequestBody @Valid Car car) {
+    return carService.createCar(car);
+}
+```
+
+Now we can add a handler to deal with MethodArgumentNotValidException exceptions, which are thrown if the validation fails. A typical @ControllerAdvice handler method would be:
```

Comment:
I think the more standard way of dealing with input validation violation is in something like the following manner:

`
    @RequestMapping(value = "/cars", method = POST)
    public RestResult<Car> createCar(@Valid CarForm carForm, Errors errors) {
        if (!errors.hasErrors()) {
            Car car = carForm.toCar(); //Or of course using a bean-mapping library of some sort ;-)
            return new RestResult<>(carService.createCar(car));
        } else {
            return new RestResult<>(errors);
        }
    }
`

In my opinion incorrect user input is not an exceptional situation either, it is something you can completely check up-front and block at that point. Exception control flow is not needed to provide the user with feedback on his/her input.